### PR TITLE
Fix composite op redispatch for NJT in inference mode

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134683
* #136566

Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

### CODE REVIEW

**Review by soulitzer — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Pretty cool!

**Comment by jbschlosser** [comment]
follow-up todo: open an issue for the wrapper subclass + enable reentrant dispatch + torch dispatch mode problem. Proposed fix is to tweak the Python key fallback impl so that it works fine when there is no mode nor subclass passed as an argument.

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -7356,6 +7356,38 @@ def g(nt):
...
+        # expect view
+        nt = random_nt_from_dims(
+            [4, None, 48],
+            device=device,
+            dtype=dtype,
+            layout=torch.jagged,
+            requires_grad=True,
+        )
+
+        with torch.inference_mode():
```
Comment (YuqingJ) [inline test/test_nestedtensor.py:7372]: just checking, this PR doesn't fix the issue when there is a custom TorchDispatchMode right?

Comment (soulitzer) [reply test/test_nestedtensor.py:7372]: that's correct

Comment (YuqingJ) [reply test/test_nestedtensor.py:7372]: Thanks. Maybe we can add a unit case test that cover the custom TorchDispatchMode in test_nestedtensor?

Comment (jbschlosser) [reply test/test_nestedtensor.py:7372]: Sure, I'll add an NJT specific test as well :)

**Pull request closed without merging**